proto: Port *.proto to proto3

wking opened the pull request:
Define [syntax](https://developers.google.com/protocol-buffers/docs/proto3#simple) to fix:

```
protoc --go_out=./go config.proto
[libprotobuf WARNING google/protobuf/compiler/parser.cc:492] No
  syntax specified for the proto file. Please use 'syntax =
  "proto2";' or 'syntax = "proto3";' to specify a syntax
  version. (Defaulted to proto2 syntax.)
```

Drop `optional` (with `sed -i 's/\toptional /\t/' *.proto`) to fix:

```
protoc --go_out=./go config.proto
config.proto:9:18: Explicit 'optional' labels are disallowed in the
  Proto3 syntax. To define 'optional' fields in Proto3, simply
  remove the 'optional' label, as fields are 'optional' by default.
```

Replace the `User` extensions with [`Any`](https://developers.google.com/protocol-buffers/docs/proto3#any) (see also,
google/protobuf#828) to fix:

```
protoc --go_out=./go config.proto
config.proto: Extensions in proto3 are only allowed for defining
  options.
```

Drop `required` (with `sed -i 's/\trequired /\t/' *.proto`) to fix:

```
protoc --go_out=./go runtime_config.proto
runtime_config.proto: Required fields are not allowed in proto3.
```

Drop `DefaultState` to fix:

```
protoc --go_out=./go runtime_config.proto
runtime_config.proto: Explicit default values are not allowed in
  proto3.
```

There's still some trouble with the resulting Go:

```
go run ./example.go
go/config.pb.go:26:8: cannot find package "google/protobuf" in any of:
        /usr/lib/go/src/google/protobuf (from $GOROOT)
        /home/wking/.local/lib/go/src/google/protobuf (from $GOPATH)
Makefile:31: recipe for target 'example' failed
```

But I haven't been able to figure that out yet.

wking committed:
proto: Port *.proto to proto3

Define syntax to fix [1]:

  protoc --go_out=./go config.proto
  [libprotobuf WARNING google/protobuf/compiler/parser.cc:492] No
    syntax specified for the proto file. Please use 'syntax =
    "proto2";' or 'syntax = "proto3";' to specify a syntax
    version. (Defaulted to proto2 syntax.)

Drop 'optional' (with 'sed -i 's/\toptional /\t/' *.proto') to fix:

  protoc --go_out=./go config.proto
  config.proto:9:18: Explicit 'optional' labels are disallowed in the
    Proto3 syntax. To define 'optional' fields in Proto3, simply
    remove the 'optional' label, as fields are 'optional' by default.

Replace the User extensions with 'Any' [2,3] to fix:

  protoc --go_out=./go config.proto
  config.proto: Extensions in proto3 are only allowed for defining
    options.

Drop required (with 'sed -i 's/\trequired /\t/' *.proto') to fix:

  protoc --go_out=./go runtime_config.proto
  runtime_config.proto: Required fields are not allowed in proto3.

Drop DefaultState to fix:

  protoc --go_out=./go runtime_config.proto
  runtime_config.proto: Explicit default values are not allowed in
    proto3.

There's still some trouble with the resulting Go:

  go run ./example.go
  go/config.pb.go:26:8: cannot find package "google/protobuf" in any of:
          /usr/lib/go/src/google/protobuf (from $GOROOT)
          /home/wking/.local/lib/go/src/google/protobuf (from $GOPATH)
  Makefile:31: recipe for target 'example' failed

But I haven't been able to figure that out yet.

[1]: https://developers.google.com/protocol-buffers/docs/proto3#simple
[2]: https://developers.google.com/protocol-buffers/docs/proto3#any
[3]: https://github.com/google/protobuf/issues/828

Signed-off-by: W. Trevor King <wking@tremily.us>

wking committed:
proto: Port example.go to proto3

Using jsonpb [1,2].

Drop the proto.String bits to avoid errors like:

  ./example.go:17: cannot use proto.String("linux") (type *string) as
    type string in field value

And use a reference to s to avoid:

  ./example.go:29: cannot use s (type oci.LinuxSpec) as type
     proto.Message in argument to marshaler.Marshal: oci.LinuxSpec
     does not implement proto.Message (ProtoMessage method has pointer
     receiver)

[1]: https://github.com/golang/protobuf/issues/44
[2]: http://godoc.org/github.com/golang/protobuf/jsonpb

Signed-off-by: W. Trevor King <wking@tremily.us>

wking commented:
I've added 8bead0f avoiding `Any` for now, and have ported `example.go` in a4d502c.  Output now looks like:

```
{
  "spec": {
    "version": "",
    "platform": {
      "os": "linux",
      "arch": "x86_64"
    },
    "process": {
      "terminal": false,
      "env": [
        "TERM=linux"
      ],
      "cwd": "/"
    },
    "hostname": ""
  }
}
```

wking committed:
proto: Add support for Process.User's Any

To avoid:

  go run ./example.go
  go/config.pb.go:26:8: cannot find package "google/protobuf" in any of:
          /usr/lib/go/src/google/protobuf (from $GOROOT)
          /home/wking/.local/lib/go/src/google/protobuf (from $GOPATH)
  Makefile:31: recipe for target 'example' failed

The Dockerfile changes will compile Google's any.proto to Go and
install it in the GOPATH.  Unfortunately, the jsonpb rendering isn't
quite right, since it's spitting out:

  "user": {
    "type_url": "oci.LinuxUser",
    "value": "qAYBsAYBuAYFuAYG"
  },

When it should be spitting out [1]:

  "user": {
    "@type": "type.googleapis.com/oci.LinuxUser",
    "uid": 1,
    "gid": 1,
    "additionalGids": [
      5,
      6
    ]
  },

[1]: https://developers.google.com/protocol-buffers/docs/proto3#json

Signed-off-by: W. Trevor King <wking@tremily.us>

wking commented:
I've figured out the ‘cannot find package "google/protobuf"’ error, and pushed da7a33f, which gives output like:

```
{
  "spec": {
    "version": "",
    "platform": {
      "os": "linux",
      "arch": "x86_64"
    },
    "process": {
      "terminal": false,
      "user": {
        "type_url": "oci.LinuxUser",
        "value": "qAYBsAYBuAYFuAYG"
      },
      "env": [
        "TERM=linux"
      ],
      "cwd": "/"
    },
    "hostname": ""
  }
}
```

It looks like golang/protobuf/jsonpb needs some more work to give us the expected:

```
  "user": {
    "@type": "type.googleapis.com/oci.LinuxUser",
    "uid": 1,
    "gid": 1,
    "additionalGids": [
      5,
      6
    ]
  },
```

we see with the C++ example in #3.

wking commented:
As an alternative to my Dockerfile changes in da7a33f, we could also use go.pedge.io/google-protobuf (see golang/protobuf#50).  I don't care either way.

bufdev commented:
@wking Note I'm trying to push to get go.pedge.io/google-protobuf just merged into golang/protobuf, I also have go.pedge.io/googleapis now, it's no fun to maintain, but I promise neither will go anywhere (and the go.pedge.io server is dead simple).

Of note, I also have https://github.com/peter-edge/go-tools/tree/master/protoc-all (go get go.pedge.io/tools/protoc-all), which I'm happy to walk anyone through, I'm going to rewrite it soon as it's become a bit of a mess but it takes care of all the import paths (and uses go.pedge.io/google-protobuf).

bufdev commented on proto/config.proto:
```diff
@@ -1,18 +1,22 @@
+	string hostname = 5;
 	// Mounts profile configuration for adding mounts to the container's
 	// filesystem.
 	repeated MountPoint mounts = 6;
```
I'm not sure if google is moving away from this, but the general protobuf standard has been to not use plural names for repeated fields, so this would be mount or mount_point.

bufdev commented on proto/config.proto:
```diff
@@ -38,18 +42,18 @@ message LinuxConfig {
+	google.protobuf.Any user = 2;
 	// Args specifies the binary and arguments for the application to
 	// execute.
 	repeated string args = 3;
```
same comment, arg

bufdev commented on proto/config.proto:
```diff
@@ -38,18 +42,18 @@ message LinuxConfig {
 message Platform {
 	// OS is the operating system.
-	optional string os = 1;
+	string os = 1;
```
since this is pretty static, and validation would be nice, it might be worth considering an enum:

```
enum OS {
  OS_NONE = 0; // I always have a zero value to denote the field not being set, proto3 requires a zero value.
  OS_DARWIN = 1;
  ...
}

enum Arch {
  ARCH_NONE = 0; // also note proto uses c++ enum name rules, so my general practice is to preface any enum names with the type name to avoid future issues.
 ...
}
```

bufdev commented on proto/config.proto:
```diff
@@ -57,50 +61,35 @@ message Process {
 	// Readonly makes the root filesystem for the container readonly before
 	// the process is executed.
-	optional bool readonly = 2;
+	bool readonly = 2;
```
might want to consider enum FileMode, but that may be getting silly

bufdev commented on proto/example.go:
```diff
@@ -3,27 +3,43 @@
+	}
+
+	marshaler := jsonpb.Marshaler{
+		EnumsAsString: true,
```
golang/protobuf just removed this https://github.com/golang/protobuf/commits/master so this won't compile (has been the bane of my docker image existence for a few days :) )

wking replied:
On Sat, Oct 24, 2015 at 10:40:33AM -0700, Peter Edge wrote:

> > -       EnumsAsString: true,
> 
> golang/protobuf just removed this
> https://github.com/golang/protobuf/commits/master so this won't
> compile (has been the bane of my docker image existence for a few
> days :) )

Heh, two days between making it the (golang/protobuf@9ebc6c4ed9) and
removing it completely (golang/protobuf@8a5d8e8b) is frankly
rediculous :p.  No wonder the Go world is a fan of vendoring, but how
folks manage this sort of thing without a library release process is a
mystery to me ;).  Anyhow, fixed in 8214db4, and thanks for pointing
it out :).

bufdev commented on proto/runtime_config.proto:
```diff
@@ -72,133 +68,133 @@ message LinuxRuntime {
+	Seccomp seccomp = 11;
 	// RootfsPropagation is the rootfs mount propagation mode for the container
-	optional string rootfs_propagation = 12;
+	string rootfs_propagation = 12;
```
same thing about enum, but I'm obviously enum-happy

bufdev commented on proto/runtime_config.proto:
```diff
@@ -206,63 +202,63 @@ message Network {
 }
 
 // Arg used for matching specific syscall arguments in Seccomp
 message Arg {
```
is this is seccomp/syscall specific, it might be better to have it as a nested message, ie this is now the type oci.Arg (same above for oci.Syscall), maybe oci.Seccomp.Syscall and oci.Seccomp.Syscall.Arg or oci.Seccomp.Arg

bufdev commented:
Just made a few comments with some thoughts/suggestions, no worries but hoping they may be helpful :)

wking committed:
proto/example: Drop EnumsAsString

It became the default with golang/protobuf@9ebc6c4e (2015-10-19) and
was removed completely with golang/protobuf@8a5d8e8b (jsonpb: Remove
Marshaler.EnumsAsString, 2015-10-21).

Signed-off-by: W. Trevor King <wking@tremily.us>

wking commented:
On Sat, Oct 24, 2015 at 10:44:59AM -0700, Peter Edge wrote:

> Just made a few comments with some thoughts/suggestions, no worries
> but hoping they may be helpful :)

They sound reasonable to me, but I'm trying to keep this branch as
minimal as possible (just get proto3+Go working as well as possible,
and see how far that gets us).  I'll leave the more peripheral
cleanups and idiomization to @vbatts and company if/when they decide
to actually go this route.

vbatts commented:
This is obsolete now. 😅
